---
title: "Identifying taxa coverage for each Repro gene"
author: "Megan Ewing"
date: "2025-09-05"
output: html_document
---

Here I seek to generate a fasta file for each 'gamete generation' gene, and pull the associated sequences from each bivalve genome. These will be used to generate gene trees to compare to the consensus tree to when calculating robinson-foulds distance.

first need

-   a gene list for each species that has the SPID & corresponding sequence number for that genome (done in `02-blastoncds` )

-   list of all of the unique SPIDs contained within all gene lists combined

-   configuration file with taxa names

we can then

### all unique SPIDs contained within all gene lists combined (with counts for how many species they are in)

```{r}
# get all files with full paths
files <- list.files("../output/genelist/", pattern = "^genelist_.*\\.csv$", full.names = TRUE)

# initialize list to store SPIDs by species
spid_lists <- list()

for (f in files) {
  # extract species name from the filename
  species <- sub("^genelist_(.*)\\.csv$", "\\1", basename(f))
  
  # read in gene lists
  df <- read.csv(f, stringsAsFactors = FALSE)
  
  #get unique SPIDs from this file
  spids <- unique(df$annot_w_GO.SPID)
  
  # store them
  spid_lists[[species]] <- spids
}

#cCombine into one vector w/ species labels
all_spids <- stack(spid_lists)

# count in how many unique species each SPID appears
library(dplyr)
spid_counts <- all_spids %>%
  distinct(values, ind) %>%           # remove duplicates within a species
  count(values, name = "species_count") %>%
  rename(SPID = values)

# preview
head(spid_counts)

```

```{bash}

mkdir ../output/extractsequences/
```

```{r}

# write spid list to csv
write.csv(spid_counts, "../output/extractsequences/uniqueSPIDlistandcounts.csv")
```

### generate fastas for each gene with sequences from applicable species extracted

```{r}
library(dplyr)
library(Biostrings)
library(stringr)

# Output folder
outdir <- "../output/genefastas/"
dir.create(outdir, showWarnings = FALSE, recursive = TRUE)

# Read all gene lists
files <- list.files("../output/genelist/", pattern = "^genelist_.*\\.csv$", full.names = TRUE)

genelists <- lapply(files, function(f) {
  species <- sub("^genelist_(.*)\\.csv$", "\\1", basename(f))
  df <- read.csv(f, stringsAsFactors = FALSE)
  df$species <- species
  df
})

genelists <- bind_rows(genelists)

# Pre-load all genome FASTAs
species_names <- unique(genelists$species)
genome_seqs <- list()

for (sp in species_names) {
  fna_path <- file.path("../data/genomes_refseq/ncbi_dataset/data", sp, "cds_from_genomic.fna")
  
  if (file.exists(fna_path)) {
    fasta <- readDNAStringSet(fna_path)
    
    # Strip everything after the first space in the header
    names(fasta) <- sub(" .*", "", names(fasta))
    
    genome_seqs[[sp]] <- fasta
  } else {
    message("Missing genome fasta for species: ", sp)
  }
}

# Loop over each SPID
spids <- unique(genelists$annot_w_GO.SPID)

for (spid in spids) {
  
  subset_df <- genelists %>%
    filter(annot_w_GO.SPID == spid)
  
  seqs_out <- DNAStringSet()
  
  for (i in seq_len(nrow(subset_df))) {
    species <- subset_df$species[i]
    seqid   <- subset_df$annot_w_GO.sseqid[i]
    
    if (!species %in% names(genome_seqs)) next
    
    fasta <- genome_seqs[[species]]
    
    # match by cleaned sequence ID
    idx <- which(names(fasta) == seqid)
    
    if (length(idx) == 1) {
      seq_name <- paste0(species, "|", seqid)
      seqs_out <- c(seqs_out, setNames(fasta[idx], seq_name))
    } else {
      message("Sequence ID not found: ", seqid, " in ", species)
    }
  }
  
  if (length(seqs_out) > 0) {
    outpath <- file.path(outdir, paste0(spid, ".fasta"))
    writeXStringSet(seqs_out, outpath, format = "fasta")
  }
}

```

preview output

```{bash}

head ../output/genefastas/A0AVF1.fasta
```
