---
title: "iqtree"
author: "Megan Ewing"
date: "2025-09-05"
output: html_document
---

using iq tree to generate consensus trees based on phyluce output & gene trees

following this [IQ tree tutorial](https://iqtree.github.io/doc/Tutorial)

will need to [install IQ tree](https://iqtree.github.io/) and adjust file path as needed for IQ tree location (I was using v3.0.1)

```{bash}
## activate mamba environment 
# eval "$(/home/shared/8TB_HDD_02/mewing0/miniforge3/bin/conda shell.bash hook)"
# eval "$(mamba shell hook --shell bash)"
#
# mamba activate
#
## create new environment in mamba for iq tree
# mamba create -n iqtree-v3.0.1 iqtree=3.0.1
#
## path to iqtree is: /home/shared/8TB_HDD_02/mewing0/miniforge3/envs/iqtree-v3.0.1/bin/iqtree3
```

to determine tree model, run model finder plus (MFP)

```{bash}
mkdir ../output/iqtree/
```

# Consensus Trees

### *50p v1 consensus*

model finder

```{bash}
/home/shared/8TB_HDD_02/mewing0/miniforge3/envs/iqtree-v3.0.1/bin/iqtree3 \
  -s ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-50p_v1-raxml/mafft-nexus-internal-trimmed-gblocks-clean-50p_v1-raxml.phylip \
  -m MFP \
  &> ../output/iqtree/50p_v1_mfp.txt

# Print the best-fit model line to console
grep "Best-fit model:" ../output/iqtree/50p_v1_mfp.txt
```

making tree, 10k bootstraps and 10k alrt iterations

```{bash}
# run iq tree with model (-m) defined by MFP
# 10000 bootstraps and 10000 alrt iterations
# gcf_000327385
# output will be saved to ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-75p_v2-raxml/

/home/shared/8TB_HDD_02/mewing0/miniforge3/envs/iqtree-v3.0.1/bin/iqtree3  \
-s ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-50p_v1-raxml/mafft-nexus-internal-trimmed-gblocks-clean-50p_v1-raxml.phylip \
-m TVM+F+R3 \
-bb 10000 \
-alrt 10000 \
-o gcf_000327385 \
-bnni \
-nt AUTO \
-redo
```

#### update tip names

back when I ran phyluce, I needed to truncate the genome names. here I will add back the suffix (.1, .2, etc) so it matches the tip names found in the gene trees.

```{bash}
# --- Define paths ---
refseq=~/bivalve_gametogenesis/data/genomes_refseq/ncbi_dataset/data
tree=~/bivalve_gametogenesis/data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-50p_v1-raxml/mafft-nexus-internal-trimmed-gblocks-clean-50p_v1-raxml.phylip.treefile
out=~/bivalve_gametogenesis/output/iqtree/50v1_fullname.treefile

# --- Build a mapping of truncated -> full (only GCF_ dirs) ---
ls -1 "$refseq" | grep '^GCF_' | while read full; do
    short=$(echo "$full" | sed 's/\.[0-9]\+$//')   # remove version suffix
    echo -e "$short\t$full"
done > mapping.tsv

# --- Copy tree to new output file ---
cp "$tree" "$out"

# --- Replace lowercase truncated names with uppercase full names ---
while IFS=$'\t' read -r short full; do
    short_lc=$(echo "$short" | tr '[:upper:]' '[:lower:]')
    sed -i "s/\b$short_lc\b/$full/g" "$out"
done < mapping.tsv

echo "updated tree written to: $out"

head $out
```

#### ggtree

```{r}
## need gg tree for visualization
# install.packages("remotes")
# remotes::install_github("YuLab-SMU/ggtree")

library(treeio)
library(tidyverse)
library(ggtree)

```

```{r}
tree <- read.tree("../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-50p_v1-raxml/mafft-nexus-internal-trimmed-gblocks-clean-50p_v1-raxml.phylip.treefile")

tree_rooted <- root(tree, outgroup = "gcf_000327385", resolve.root = TRUE)

p <- ggtree(tree_rooted, layout = "rectangular") +
     geom_tiplab(size = 3, fontface = "italic") +
     geom_text2(aes(subset = !isTip, label = label), size = 2.5, vjust = -0.3) +
     theme_tree2()

# Save as PNG
ggsave(
  filename = "50v1_tree.png", 
  plot = p,                      
  width = 8, height = 8,        
  dpi = 300,                     
)
```

### *50p v2 consensus*

mfp

```{bash}
/home/shared/8TB_HDD_02/mewing0/miniforge3/envs/iqtree-v3.0.1/bin/iqtree3 \
  -s ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-50p_v2-raxml/mafft-nexus-internal-trimmed-gblocks-clean-50p_v2-raxml.phylip \
  -m MFP \
  &> ../output/iqtree/50p_v2_mfp.txt

# Print the best-fit model line to console
grep "Best-fit model:" ../output/iqtree/50p_v2_mfp.txt
```

making tree, 10k bootstraps and 10k alrt iterations

```{bash}
# run iq tree with model (-m) defined by MFP
# 10000 bootstraps and 10000 alrt iterations
# gcf_000327385
# output will be saved to ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-75p_v2-raxml/

/home/shared/8TB_HDD_02/mewing0/miniforge3/envs/iqtree-v3.0.1/bin/iqtree3  \
-s ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-50p_v2-raxml/mafft-nexus-internal-trimmed-gblocks-clean-50p_v2-raxml.phylip \
-m GTR+F+I+R3 \
-bb 10000 \
-alrt 10000 \
-o gcf_000327385 \
-bnni \
-nt AUTO \
-redo
```

### *75p v1 consensus*

```{bash}

/home/shared/8TB_HDD_02/mewing0/miniforge3/envs/iqtree-v3.0.1/bin/iqtree3 \
  -s ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-75p_v1-raxml/mafft-nexus-internal-trimmed-gblocks-clean-75p_v1-raxml.phylip \
  -m MFP \
  &> ../output/iqtree/75_v1_mfp.txt

# Print the best-fit model line to console
grep "Best-fit model:" ../output/iqtree/75_v1_mfp.txt

```

```{bash}
# run iq tree with model (-m) defined by MFP
# 10000 bootstraps and 10000 alrt iterations
# gcf_000327385
# output will be saved to ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-75p_v2-raxml/

/home/shared/8TB_HDD_02/mewing0/miniforge3/envs/iqtree-v3.0.1/bin/iqtree3  \
-s ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-75p_v1-raxml/mafft-nexus-internal-trimmed-gblocks-clean-75p_v1-raxml.phylip \
-m TVM+F+I+R3 \
-bb 10000 \
-alrt 10000 \
-o gcf_000327385 \
-bnni \
-nt AUTO \
-redo
```

### *75p v2 consensus*

```{bash}

/home/shared/8TB_HDD_02/mewing0/miniforge3/envs/iqtree-v3.0.1/bin/iqtree3 \
  -s ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-75p_v2-raxml/mafft-nexus-internal-trimmed-gblocks-clean-75p_v2-raxml.phylip \
  -m MFP \
  &> ../output/iqtree/75_v2_mfp.txt

# Print the best-fit model line to console
grep "Best-fit model:" ../output/iqtree/75_v2_mfp.txt

```

```{bash}
# run iq tree with model (-m) defined by MFP
# 10000 bootstraps and 10000 alrt iterations
# gcf_000327385
# output will be saved to ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-75p_v2-raxml/

/home/shared/8TB_HDD_02/mewing0/miniforge3/envs/iqtree-v3.0.1/bin/iqtree3  \
-s ../data/phyluce/taxon-set1/mafft-nexus-internal-trimmed-gblocks-clean-75p_v2-raxml/mafft-nexus-internal-trimmed-gblocks-clean-75p_v2-raxml.phylip \
-m GTR+F+I+R3 \
-bb 10000 \
-alrt 10000 \
-o gcf_000327385 \
-bnni \
-nt AUTO \
-redo
```

# Gene Trees

```{bash}

# Paths to binaries
IQTREE=/home/shared/8TB_HDD_02/mewing0/miniforge3/envs/iqtree-v3.0.1/bin/iqtree3
MAFFT=/usr/bin/mafft        # adjust if MAFFT is elsewhere

FASTA_DIR="../output/genefastas"
IQTREE_DIR="../output/iqtree"
THREADS=4                  # number of genes to run simultaneously
LOG_FILE="$IQTREE_DIR/genetree.log"

# clear previous log
> "$LOG_FILE"

run_iqtree() {
    fasta="$1"
    gene=$(basename "$fasta" .fasta)
    outdir="$IQTREE_DIR/$gene"
    mkdir -p "$outdir"

    echo "[$(date '+%Y-%m-%d %H:%M:%S')] Starting $gene" >> "$LOG_FILE"

    # Step 0: Align sequences with MAFFT
    aligned_fasta="$outdir/${gene}_aligned.fasta"
    echo "[$(date '+%Y-%m-%d %H:%M:%S')] Aligning $gene with MAFFT ..." >> "$LOG_FILE"
    $MAFFT --auto "$fasta" > "$aligned_fasta"

    # Step 1: Model Finder Plus
    echo "[$(date '+%Y-%m-%d %H:%M:%S')] Running Model Finder Plus for $gene ..." >> "$LOG_FILE"
    $IQTREE -s "$aligned_fasta" -m MFP -pre "$outdir/$gene"

    # Extract best-fit model from the .iqtree file
    model=$(grep "Best-fit model:" "$outdir/$gene.iqtree" | awk '{print $3}')
    echo "[$(date '+%Y-%m-%d %H:%M:%S')] Best-fit model for $gene: $model" >> "$LOG_FILE"

    # Step 2: Generate tree
    echo "[$(date '+%Y-%m-%d %H:%M:%S')] Generating tree for $gene ..." >> "$LOG_FILE"
    $IQTREE \
        -s "$aligned_fasta" \
        -m "$model" \
        -bb 10000 \
        -alrt 10000 \
        -o gcf_000327385 \
        -bnni \
        -nt AUTO \
        -redo \
        -pre "$outdir/$gene"

    echo "[$(date '+%Y-%m-%d %H:%M:%S')] Finished $gene" >> "$LOG_FILE"
}

export -f run_iqtree
export IQTREE MAFFT IQTREE_DIR LOG_FILE

# run in parallel
find "$FASTA_DIR" -name "*.fasta" | parallel -j $THREADS run_iqtree {}


```

```{bash}
cd ../output/genefastas/
# total number of files or subdir in source directory 
ls -1 | wc -l


cd ../iqtree/
# total number of files or subdir in created directory (should equal above + consesus tree files + log file)
ls -1 | wc -l

# number of .treefiles created 
find . -type f -name "*.treefile" | wc -l

```

**note**: the number of .treefiles created is less than the total number of fastas given to iqtree -- why? IQ tree ignores any fastas (inputs) with less than 3 sequences. which makes sense since a tree with two sequences would just look like: **[**

we can confirm this by comparing the missing tree file with the SPID counts \< 3

```{r}
# create a dataframe from the input SPID list where counts (# of species represented) are less than 3

spidcounts <- read_csv("../output/extractsequences/uniqueSPIDlistandcounts.csv")
head(spidcounts)

shortseqs <- spidcounts[spidcounts$species_count < 3, "SPID"]
head(shortseqs)

```

```{r}

# find which of the iqtree subdirectories do not contain tree files, save as dataframe `shorttrees`

# list subdirectories (non-recursive)
SPID <- list.dirs(path = "../output/iqtree/", full.names = TRUE, recursive = FALSE)

# check each for .treefile
missing_treefile <- character()
for (d in SPID) {
  treefiles <- list.files(d, pattern = "\\.treefile$", full.names = TRUE)
  if (length(treefiles) == 0) {
    missing_treefile <- c(missing_treefile, d)
  }
}

missing_treefile

shorttrees <- data.frame(
  SPID = basename(missing_treefile),
  stringsAsFactors = FALSE
)

shorttrees

```

compare input files \<3 to the iqtree subdirs missing tree files. if returns TRUE, the only iqtrees that weren't made were the ones with less than 3 sequences.

```{r}
# library(dplyr)

#checks if the lists of 
all_equal(shortseqs, shorttrees)

```

yay! TRUE returned. can now move on and start comparing our gene trees and chosen consensus tree.

we will also be interested in these genes independently though, as they have low representation across our taxa!
